File: autobyteus-android/app/src/main/java/org/autobyteus/mobile/connection/PairingLinkParser.kt
package org.autobyteus.mobile.connection

import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Base64

data class PairingPayload(
    val serverBaseUrl: String,
    val pairingCode: String,
    val serverName: String?,
)

data class ParsedPairingInput(
    val profile: SavedNodeProfile,
    val webViewUrl: String,
    val hasPairingPayload: Boolean,
)

object PairingLinkParser {
    fun parse(rawText: String): ParsedPairingInput {
        val text = rawText.trim()
        if (text.isEmpty()) {
            throw NodeUrlNormalizationException("Paste a Phone Access link, pairing payload, or node URL.")
        }

        val urlCandidate = if (text.startsWith("{")) null else extractUrl(text)
        if (urlCandidate != null) {
            return parseUrl(urlCandidate)
        }

        val payload = try {
            parsePairingPayload(text)
        } catch (_: NodeUrlNormalizationException) {
            val normalized = NodeUrlNormalizer.normalize(text)
            return ParsedPairingInput(
                profile = SavedNodeProfile.fromNormalized(normalized),
                webViewUrl = normalized.mobileUrl,
                hasPairingPayload = false,
            )
        }
        val encoded = encodePairingPayload(text, payload)
        val normalized = NodeUrlNormalizer.normalize(payload.serverBaseUrl)
        val profile = SavedNodeProfile.fromNormalized(
            normalized,
            displayName = payload.serverName,
            httpAcknowledged = normalized.scheme == "https",
        )
        return ParsedPairingInput(
            profile = profile,
            webViewUrl = "${normalized.mobileUrl}?pairing=$encoded",
            hasPairingPayload = true,
        )
    }

    private fun parseUrl(url: String): ParsedPairingInput {
        val uri = URI(url)
        val query = parseQuery(uri.rawQuery)
        val pairingParam = query["pairing"]
        if (!pairingParam.isNullOrBlank()) {
            val payload = parsePairingPayload(pairingParam)
            val normalized = NodeUrlNormalizer.normalize(payload.serverBaseUrl)
            val profile = SavedNodeProfile.fromNormalized(
                normalized,
                displayName = payload.serverName,
                httpAcknowledged = normalized.scheme == "https",
            )
            return ParsedPairingInput(
                profile = profile,
                webViewUrl = "${normalized.mobileUrl}?pairing=$pairingParam",
                hasPairingPayload = true,
            )
        }

        val normalized = NodeUrlNormalizer.normalize(url)
        return ParsedPairingInput(
            profile = SavedNodeProfile.fromNormalized(normalized),
            webViewUrl = normalized.mobileUrl,
            hasPairingPayload = false,
        )
    }

    fun parsePairingPayload(rawPayload: String): PairingPayload {
        val trimmed = rawPayload.trim()
        val json = when {
            trimmed.startsWith("{") -> trimmed
            else -> decodeBase64Url(trimmed)
        }
        val serverBaseUrl = extractJsonString(json, "serverBaseUrl")
            ?: throw NodeUrlNormalizationException("Pairing payload is missing serverBaseUrl.")
        val pairingCode = extractJsonString(json, "pairingCode")
            ?: throw NodeUrlNormalizationException("Pairing payload is missing pairingCode.")
        val serverName = extractJsonString(json, "serverName")
        return PairingPayload(serverBaseUrl, pairingCode, serverName)
    }

    private fun encodePairingPayload(raw: String, payload: PairingPayload): String {
        val trimmed = raw.trim()
        if (!trimmed.startsWith("{")) {
            return trimmed
        }
        return Base64.getUrlEncoder()
            .withoutPadding()
            .encodeToString(trimmed.toByteArray(StandardCharsets.UTF_8))
    }

    private fun decodeBase64Url(value: String): String {
        return try {
            String(Base64.getUrlDecoder().decode(value), StandardCharsets.UTF_8)
        } catch (error: IllegalArgumentException) {
            throw NodeUrlNormalizationException("Pairing payload must be a Phone Access URL, base64url payload, or JSON payload.")
        }
    }

    private fun parseQuery(rawQuery: String?): Map<String, String> {
        if (rawQuery.isNullOrBlank()) {
            return emptyMap()
        }
        return rawQuery.split('&')
            .mapNotNull { part ->
                val idx = part.indexOf('=')
                if (idx < 0) return@mapNotNull null
                val key = URLDecoder.decode(part.substring(0, idx), StandardCharsets.UTF_8.name())
                val value = URLDecoder.decode(part.substring(idx + 1), StandardCharsets.UTF_8.name())
                key to value
            }
            .toMap()
    }

    private fun extractUrl(text: String): String? {
        val direct = text.takeIf { URL_PREFIX.containsMatchIn(it) }
        if (direct != null) {
            return direct
        }
        return URL_IN_TEXT.find(text)?.value
    }

    private fun extractJsonString(json: String, key: String): String? {
        val regex = Regex("\\\"${Regex.escape(key)}\\\"\\s*:\\s*\\\"((?:\\\\.|[^\\\"])*)\\\"")
        val match = regex.find(json) ?: return null
        return unescapeJsonString(match.groupValues[1])
    }

    private fun unescapeJsonString(value: String): String = buildString {
        var index = 0
        while (index < value.length) {
            val char = value[index]
            if (char != '\\' || index == value.lastIndex) {
                append(char)
                index += 1
                continue
            }
            val escaped = value[index + 1]
            when (escaped) {
                '"', '\\', '/' -> append(escaped)
                'b' -> append('\b')
                'f' -> append('\u000C')
                'n' -> append('\n')
                'r' -> append('\r')
                't' -> append('\t')
                'u' -> {
                    val hex = value.substring(index + 2, index + 6)
                    append(hex.toInt(16).toChar())
                    index += 4
                }
                else -> append(escaped)
            }
            index += 2
        }
    }

    private val URL_PREFIX = Regex("^https?://", RegexOption.IGNORE_CASE)
    private val URL_IN_TEXT = Regex("https?://[^\\s<>\"']+", RegexOption.IGNORE_CASE)
}

File: autobyteus-ios/AutoByteusMobileCore/PairingLinkParser.swift
import Foundation

public struct PairingPayload: Codable, Equatable {
    public let serverBaseURL: String
    public let pairingCode: String
    public let serverName: String?

    public init(serverBaseURL: String, pairingCode: String, serverName: String?) {
        self.serverBaseURL = serverBaseURL
        self.pairingCode = pairingCode
        self.serverName = serverName
    }

    private enum CodingKeys: String, CodingKey {
        case serverBaseURL = "serverBaseUrl"
        case pairingCode
        case serverName
    }
}

public struct ParsedPairingInput: Equatable {
    public let profile: SavedNodeProfile
    public let webViewURL: String
    public let hasPairingPayload: Bool

    public init(profile: SavedNodeProfile, webViewURL: String, hasPairingPayload: Bool) {
        self.profile = profile
        self.webViewURL = webViewURL
        self.hasPairingPayload = hasPairingPayload
    }
}

public enum PairingLinkParser {
    public static func parse(_ rawText: String) throws -> ParsedPairingInput {
        let text = rawText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else {
            throw NodeURLNormalizationError("Paste a Phone Access link, pairing payload, or node URL.")
        }

        if !text.hasPrefix("{"), let urlCandidate = extractURL(from: text) {
            return try parseURL(urlCandidate)
        }

        do {
            let payload = try parsePairingPayload(text)
            let encoded = try encodePairingPayload(raw: text, payload: payload)
            let normalized = try NodeURLNormalizer.normalize(payload.serverBaseURL)
            let profile = SavedNodeProfile.fromNormalized(
                normalized,
                displayName: payload.serverName,
                httpAcknowledged: normalized.scheme == "https"
            )
            return ParsedPairingInput(
                profile: profile,
                webViewURL: "\(normalized.mobileURL)?pairing=\(encoded)",
                hasPairingPayload: true
            )
        } catch let payloadError as NodeURLNormalizationError {
            do {
                let normalized = try NodeURLNormalizer.normalize(text)
                return ParsedPairingInput(
                    profile: SavedNodeProfile.fromNormalized(normalized),
                    webViewURL: normalized.mobileURL,
                    hasPairingPayload: false
                )
            } catch {
                throw payloadError
            }
        }
    }

    public static func parsePairingPayload(_ rawPayload: String) throws -> PairingPayload {
        let trimmed = rawPayload.trimmingCharacters(in: .whitespacesAndNewlines)
        let data: Data
        if trimmed.hasPrefix("{") {
            guard let jsonData = trimmed.data(using: .utf8) else {
                throw NodeURLNormalizationError("Pairing payload JSON must be valid UTF-8.")
            }
            data = jsonData
        } else {
            data = try decodeBase64URL(trimmed)
        }

        do {
            let payload = try JSONDecoder().decode(PairingPayload.self, from: data)
            guard !payload.serverBaseURL.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw NodeURLNormalizationError("Pairing payload is missing serverBaseUrl.")
            }
            guard !payload.pairingCode.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw NodeURLNormalizationError("Pairing payload is missing pairingCode.")
            }
            return payload
        } catch let normalizerError as NodeURLNormalizationError {
            throw normalizerError
        } catch {
            throw NodeURLNormalizationError("Pairing payload must include serverBaseUrl and pairingCode.")
        }
    }

    private static func parseURL(_ url: String) throws -> ParsedPairingInput {
        guard let components = URLComponents(string: url) else {
            throw NodeURLNormalizationError("The URL could not be parsed.")
        }
        let pairingParam = components.queryItems?.first { $0.name == "pairing" }?.value
        if let pairingParam, !pairingParam.isEmpty {
            let payload = try parsePairingPayload(pairingParam)
            let normalized = try NodeURLNormalizer.normalize(payload.serverBaseURL)
            let profile = SavedNodeProfile.fromNormalized(
                normalized,
                displayName: payload.serverName,
                httpAcknowledged: normalized.scheme == "https"
            )
            return ParsedPairingInput(
                profile: profile,
                webViewURL: "\(normalized.mobileURL)?pairing=\(pairingParam)",
                hasPairingPayload: true
            )
        }

        let normalized = try NodeURLNormalizer.normalize(url)
        return ParsedPairingInput(
            profile: SavedNodeProfile.fromNormalized(normalized),
            webViewURL: normalized.mobileURL,
            hasPairingPayload: false
        )
    }

    private static func encodePairingPayload(raw: String, payload: PairingPayload) throws -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.hasPrefix("{") else { return trimmed }
        let data = try JSONEncoder().encode(payload)
        return data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func decodeBase64URL(_ value: String) throws -> Data {
        var base64 = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        guard let data = Data(base64Encoded: base64) else {
            throw NodeURLNormalizationError("Pairing payload must be a Phone Access URL, base64url payload, or JSON payload.")
        }
        return data
    }

    private static func extractURL(from text: String) -> String? {
        if text.range(of: #"^https?://"#, options: [.regularExpression, .caseInsensitive]) != nil {
            return text
        }
        guard let range = text.range(
            of: #"https?://[^\s<>"']+"#,
            options: [.regularExpression, .caseInsensitive]
        ) else {
            return nil
        }
        return String(text[range])
    }
}
